typescript-for-haskellers/article.md

```diff
@@ -0,0 +1,1002 @@
+The syntax looks like `[set of field names]: type;`.
+For example, you can add optional modifier or readonly modifier to each field with `type Partial<T> = { [P in keyof T]?: T[P] }` and `type Readonly<T> = { readonly [P in keyof T]: T[P] }`.
+It is also possible to remove such modifiers using `-?` and `-readonly`.
+
+The `keyof` operator in the code below produces a union of properties name.
+Square brackets indicates computed names.
+And `in` operator goes over all field keys.
+
+`Partial` type adds `?` modifier for each property.
+
+```typescript
+type Partial<T> = { [P in keyof T]?: T[P] };
+
+type A = { x: number; y: number };
+type PartialA = Partia<A>;
+
+const a: A = { x: 1 }; // Property 'y' is missing in type '{ x: number; }' but required in type 'A'.
+const a: PartialA = { x: 1 }; // Ok
+```
+
+Let's try now to remove `?` and `readonly` modifier from all properties.
+
+```typescript
+type NoModifiers<T> = { -readonly [P in keyof T]-?: T[P] };
+
+type A = { readonly x?: number; readonly y?: number };
+type NoModifiersA = NoModifiers<A>;
+
+const a: A = { x: 1 }; // Ok
+a.x = 1; // Cannot assign to 'x' because it is a read-only property.
+
+const a: NoModifiersA = { x: 1 }; // Property 'y' is missing in type '{ x: number; }' but required in type 'NoModifiers<A>'
+const a: NoModifiersA = { x: 1, y: 2 }; // Ok
+a.x = 2; // Ok
+```
+
+Let's present some complex examples of mapped types usage.
+
+Defining type which contains string and numerical properties.
+
+```typescript
+type Point = {
+  x: number;
+  y: number;
+  0: string;
+}
+```
+
+Now, using mapped types we will create a type which will pick only some properties.
+Word `extends` here say that generic `K` will contain a subset of union properties keys `T`.
+And the resulting type will contain only properties from `K`.
+
+```typescript
+type Pick<T, K extends keyof T> = {
+  [P in K]: T[P];
+};
+
+type PointValues = Pick<Point, "x" | 0> // type PointValues = { x: number; 0: number }
+```
+
+The next example will be more complex.
+We will try to create the type which produce type with getters for all string properties.
+
+To implement it lets firstly define helper type `Getter`.
+Capitalize is a build in type which may be applied for the string type.
+So, for `T` which extends string we can produce getter specific name.
+
+```typescript
+type Getter<T extends string> = `get${Capitalize<T>}`;
+```
+
+And finally we can implement `Getters` type.
+Type `Extract` extract from type all union members which are assignable to some type.
+So, here for each property name `K` in `T` which is assignable to `string` we create getter property name with type `() => T[K]`.
+
+```typescript
+type Getters<T> = {
+    [K in keyof T as Getter<Extract<K, string>>]: () => T[K];
+}
+
+type PointGetters = Getters<Point> // type PointValues = { getX: () => number; getY: () => number; }
+```
+
+#### Conditional types
+
+Another great feature is conditional types.
```

Comment:
Structure is a bit off. 

I would just:

give a definition
show basic example 
tell that it can be used in conjunction with mapped types for that and that 
give a definition of distributed conditional types
show an example 

so that you are not jumping around stuff